Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/NativeLibrary.java

```diff
@@ -401,18 +406,69 @@ private NativeLibrary()
 		CacheClassesAndMethods();
 	}
 
-	public static void displayAlertMsg(final String alert)
+	private static boolean alertResult = false;
+	public static boolean displayAlertMsg(final String caption, final String text, final boolean yesNo)
 	{
-		Log.error("[NativeLibrary] Alert: " + alert);
+		Log.error("[NativeLibrary] Alert: " + text);
 		final EmulationActivity emulationActivity = sEmulationActivity.get();
-		if (emulationActivity != null)
+		boolean result = false;
+		if (emulationActivity == null)
 		{
-			emulationActivity.runOnUiThread(() -> Toast.makeText(emulationActivity, "Panic Alert: " + alert, Toast.LENGTH_LONG).show());
+			Log.warning("[NativeLibrary] EmulationActivity is null, can't do panic alert.");
 		}
 		else
 		{
-			Log.warning("[NativeLibrary] EmulationActivity is null, can't do panic toast.");
+			// Create a Handler, which allows us to effectively block whilst waiting for the user's input,
+			// then throws on completion, allowing us to exit the loop.
+			Looper.prepare();
+			final Handler handler = new Handler() {
+				@Override
+				public void handleMessage(Message mesg) {
+					throw new RuntimeException();
```

Comment:
I tried implementing it that way, but unfortunately it does not work, probably because Loopers aren't really designed to be used this way. They're meant to be just fire-and-forget in a thread, (infact ‎`quit()` ['exists mostly for historical and testing reasons'](https://groups.google.com/forum/#!topic/android-platform/_jsUqXi6DKM)), and seemingly the only way to actually exit from the `loop()` call is by throwing. The whole problem is that the call for a yes/no dialog from the dolphin C++ code is a single synchronous call that expects the response as the return value, whereas in Android the dialogs, and indeed many other things, are asynchronous. Unfortunately I think any solution other than changing the base C++, which is more than I think I could do, is likely to be hacky.

Reply:
I also think that this is why the dialogs weren't implemented in the first place, as it is a tricky problem.

Reply:
fair enough, I will not stop this from being merge but my personal thought is that it looks like Looper is not what we should use here. The traditional Object.wait()/Object.notify() are enough and less hack :)